JS: add query js/unsafe-jquery-plugin

esbena opened the pull request:
This is the production quality version of <https://securitylab.github.com/ctf/jquery>.

The query has the same main ideas as the one in the competition, but it has been amended with lots of type tracking and advanced sanitization.

Evaluation is ongoing.

erik-krogh commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  /**
+   * Holds if `plugin` has a default option defined a `def`.
```
```suggestion
   * Holds if `plugin` has a default option defined at `def`.
```

erik-krogh commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,242 @@
+    IsElementSanitizer() {
+      // common ad hoc sanitizing calls
+      exists(string name | getCalleeName() = name |
+        name = "isElement" or name = "isWindow" or name = "isWindow"
```
```suggestion
        name = "isElement" or name = "isWindow"
```

erik-krogh commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  /**
+   * Expression of form `typeof x.jquery !== "undefined"` or `x.jquery`, which sanitizes `x`.
```
It also matches expressions of the form `x.jquery !== undefined`, maybe that case should also be mentioned in the docstring.

esbena commented:
I have addressed the comments and improved the sanitizers a bit. 
The evaluation is uneventful for [big-apps](https://git.semmle.com/esben/dist-compare-reports/tree/js/unsafe-jquery_1579343789129).

asgerf commented on javascript/ql/src/Security/CWE-079/UnsafeJQueryPlugin.qhelp:
```diff
@@ -0,0 +1,98 @@
+	<recommendation>
+		<p>
+
+			Document all options that can lead to cross-site scripting attacks.
```
... or guard against unsafe inputs.

asgerf commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPlugin.qll:
```diff
@@ -0,0 +1,55 @@
+      // prefixing prevents forced html/css confusion:
+
+      // prefixing through concatenation:
+      succ.asExpr().(AddExpr).getRightOperand().flow() = pred
```
Is there a reason not to use `StringConcatenation` here?

asgerf commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,244 @@
+  ) {
+    t.start() and
+    jQueryPluginDefinition(_, def) and
+    result.flowsTo(def)
```
The return type should generally be `SourceNode` in a recursive type tracking predicate.

With the current return type of `FunctionNode` we won't track any further than `getALocalSource()` since e.g. a `ParameterNode` or `PropRead` can't be included in the base case of the predicate.

A better return type can be set on the "end" predicate, which I'd recommend always defining explicitly.

asgerf commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,244 @@
+    exists(ExtendCall extend, JQueryPluginOptions options, DataFlow::SourceNode default |
+      options.getPlugin() = plugin and
+      options = getAnExtendOperand(DataFlow::TypeBackTracker::end(), extend) and
+      default = getAnExtendOperand(DataFlow::TypeBackTracker::end(), extend) and
```
Could you factor out the "end" predicate of `getAnExtendOperand`?

asgerf commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll:
```diff
@@ -0,0 +1,244 @@
+  /**
+   * Expression like `typeof x.<?> !== "undefined"` or `x.<?>`, which sanitizes `x`, as it is unlikely to be a string afterwards.
+   */
+  class PropertyPrecenseSanitizer extends TaintTracking::SanitizerGuardNode, DataFlow::ValueNode {
```
`Precense` -> `Presence`

esbena commented:
All comments have been addressed.

mchammer01 commented:
I'll take a look at this shortly.

mchammer01 commented on javascript/ql/src/Security/CWE-079/UnsafeJQueryPlugin.qhelp:
```diff
@@ -0,0 +1,100 @@
+
+		<p>
+
+			This is however not a safe plugin, since the call to
```
```suggestion
			This is, however, not a safe plugin, since the call to
```

asgerf commented:
Thanks for the update. The query looks great! LGTM from my end, just waiting for the evaluation.

esbena commented:
Thanks @mchammer01, all of your suggestions were clear and have been implemented.

I will produce one last evaluation of this query in the weekend.

mchammer01 commented:
Thanks for the documentation updates @esbena!

esbena commented:
I am happy with the performance now.

Initial evaluation after the review suggestions had three timeouts: https://git.semmle.com/esben/dist-compare-reports/tree/js/unsafe-jquery_1580568281956

They have been fixed in e118049 (worth a look): https://git.semmle.com/esben/dist-compare-reports/tree/js/unsafe-jquery_1580804769115.

asgerf commented on javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPlugin.qll:
```diff
@@ -36,7 +36,7 @@ module UnsafeJQueryPlugin {
 
       // prefixing through concatenation:
-      StringConcatenation::getFirstOperand(succ) != pred
+      StringConcatenation::getOperand(succ, [1..StringConcatenation::getNumOperand(succ) - 1]) = pred
```
👍  quite happy to see this use of `!=` go away, regardless of performance.

Maybe it would be better to use the `taintStep` predicate exposed by the library, though?
```
StringConcatenation::taintStep(src, dst, _, any(int i | i >= 1))
```

esbena replied:
Yeah, that seems cleaner. Let me do a quick sanity check of the resulting performance..

esbena replied:
https://git.semmle.com/esben/dist-compare-reports/tree/js/unsafe-jquery_1580812266407 the reformulation does not time out, so it is probably good to go.
